File: Windows/Engine/KGArcShape.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Drawing2D;

namespace Kinogrida
{
    class KGArcShape : KGBaseShape
    {
        private float _angleOffset;
        private float _arc;
        private bool  _clockwise;
        private float _rotationAmount;

        private const float RAD2DEG = 180f / MathF.PI;

        public KGArcShape(KGGrid grid, int x, int y, Color color)
            : base(grid, x, y, color)
        {
            _angleOffset    = RandomAngleOffset();
            _arc            = 2f;
            _clockwise      = true;
            _rotationAmount = 0.25f;
        }

        private static float RandomAngleOffset()
        {
            float[] angles = { 0f, 0.25f, 0.5f, 0.75f };
            return angles[Random.Shared.Next(4)] * MathF.PI * 2f;
        }

        private struct ArcGeo
        {
            public float CenterX, CenterY;
            public float EndAngle, InitAngle;
            public float StartAngle, TailAngle;  // head and tail arc positions
            public float ArcRadius, ArcCenterX, ArcCenterY;
            public float CellSize, LineWidth;
            public float RotationDirection;
            public float OffsetX, OffsetY;
        }

        private ArcGeo ComputeGeometry(KGGridConfig config, float angleOffset, float arc)
        {
            float cs  = config.CellSize;
            float ox  = config.OffsetX, oy = config.OffsetY;
            float lw  = config.LineWidth;

            float centerX = ox + X * cs + cs / 2f;
            float centerY = oy + Y * cs + cs / 2f;

            float endAngle  = angleOffset;
            float initAngle = endAngle - MathF.PI;
            float rotDir    = _clockwise ? 1f : -1f;
            float rotAngle  = 2f * MathF.PI * _rotationAmount;

            float startAngle = initAngle + rotAngle * Progress     * rotDir;
            float tailAngle  = initAngle + rotAngle * TailProgress * rotDir;

            float arcRadius  = cs * arc;
            float arcCenterX = centerX + cs * arc * MathF.Cos(endAngle);
            float arcCenterY = centerY + cs * arc * MathF.Sin(endAngle);

            return new ArcGeo
            {
                CenterX = centerX,      CenterY = centerY,
                EndAngle = endAngle,    InitAngle = initAngle,
                StartAngle = startAngle, TailAngle = tailAngle,
                ArcRadius = arcRadius,
                ArcCenterX = arcCenterX, ArcCenterY = arcCenterY,
                CellSize = cs,           LineWidth = lw,
                RotationDirection = rotDir,
                OffsetX = ox,            OffsetY = oy
            };
        }

        private ArcGeo ComputeGeometry(KGGridConfig config) => ComputeGeometry(config, _angleOffset, _arc);

        private void DrawArcElements(Graphics g, ArcGeo geo, Color color, float width)
        {
            float halfW    = width / 2f;
            float rotAngle = 2f * MathF.PI * _rotationAmount;

            // Arc stroke spans from tailAngle (tail) to startAngle (head)
            float arcStartDeg = geo.TailAngle  * RAD2DEG;
            float arcSweepDeg = (geo.StartAngle - geo.TailAngle) * RAD2DEG;

            // Head cap center
            float hx = geo.ArcCenterX + geo.ArcRadius * MathF.Cos(geo.StartAngle);
            float hy = geo.ArcCenterY + geo.ArcRadius * MathF.Sin(geo.StartAngle);

            // Tail cap center
            float tx = geo.ArcCenterX + geo.ArcRadius * MathF.Cos(geo.TailAngle);
            float ty = geo.ArcCenterY + geo.ArcRadius * MathF.Sin(geo.TailAngle);

            using var pen = new Pen(color, geo.LineWidth) { StartCap = LineCap.Round, EndCap = LineCap.Round };

            // Outer arc
            float outerR = geo.ArcRadius + halfW;
            if (outerR > 0)
                g.DrawArc(pen,
                    geo.ArcCenterX - outerR, geo.ArcCenterY - outerR, outerR * 2f, outerR * 2f,
                    arcStartDeg, arcSweepDeg);

            // Inner arc
            float innerR = MathF.Max(0.001f, geo.ArcRadius - halfW);
            g.DrawArc(pen,
                geo.ArcCenterX - innerR, geo.ArcCenterY - innerR, innerR * 2f, innerR * 2f,
                arcStartDeg, arcSweepDeg);

            if (halfW <= 0) return;

            // Head cap (semicircle at leading edge)
            float headCapStart = (_clockwise ? geo.StartAngle : geo.StartAngle - MathF.PI) * RAD2DEG;
            g.DrawArc(pen, hx - halfW, hy - halfW, halfW * 2f, halfW * 2f, headCapStart, 180f);

            // Tail cap (semicircle at trailing edge)
            float tailCapStart = (_clockwise ? geo.TailAngle - MathF.PI : geo.TailAngle) * RAD2DEG;
            g.DrawArc(pen, tx - halfW, ty - halfW, halfW * 2f, halfW * 2f, tailCapStart, 180f);
        }

        public override void Draw(Graphics g, KGGridConfig config)
        {
            var geo      = ComputeGeometry(config);
            float baseSize = config.CellSize * 0.8f;
            DrawArcElements(g, geo, Color, baseSize);
            DrawArcElements(g, geo, System.Drawing.Color.White, baseSize - geo.LineWidth * 4f);
        }

        public override void UpdatePosition(KGGrid grid)
        {
            if (!IsMoving) return;

            if (Progress < 1f)
            {
                Progress = MathF.Min(1f, Progress + Speed);
                if (Progress >= 1f) HasReachedTarget = true;
            }
            else if (TailProgress < 1f)
            {
                TailProgress = MathF.Min(1f, TailProgress + Speed);
            }
        }

        // Called by engine after all shapes update to finalize arc position
        public void FinalizeProgress(KGGridConfig config, KGGrid grid)
        {
            if (!IsMoving || TailProgress < 1f) return;

            var geo      = ComputeGeometry(config);
            float rotAngle   = 2f * MathF.PI * _rotationAmount;
            float finalAngle = geo.EndAngle - MathF.PI + rotAngle * geo.RotationDirection;

            int newX = (int)MathF.Floor((geo.ArcCenterX + geo.ArcRadius * MathF.Cos(finalAngle) - geo.OffsetX) / geo.CellSize);
            int newY = (int)MathF.Floor((geo.ArcCenterY + geo.ArcRadius * MathF.Sin(finalAngle) - geo.OffsetY) / geo.CellSize);

            OnMoveComplete(grid, newX, newY);
            Progress = TailProgress = 0f;
        }

        public override void CalculateNewTarget(KGGrid grid, KGGridConfig config)
        {
            if (IsMoving) return;

            int maxArc = Math.Min(config.NbrColumns, config.NbrRows) / 2 - 1;
            if (maxArc < 1) return;

            float newOffset = RandomAngleOffset();
            float newArc    = Random.Shared.Next(1, maxArc + 1);
            bool  newCW     = Random.Shared.Next(2) == 0;
            float[] rotAmounts = { 0.25f, 0.5f, 0.75f };
            float newRot    = rotAmounts[Random.Shared.Next(3)];

            // Save current params
            float savedOffset = _angleOffset, savedArc = _arc;
            bool  savedCW     = _clockwise;
            float savedRot    = _rotationAmount;

            // Temporarily apply candidate params to calculate endpoint
            _angleOffset    = newOffset;
            _arc            = newArc;
            _clockwise      = newCW;
            _rotationAmount = newRot;

            var geo      = ComputeGeometry(config);
            float rotAngle   = 2f * MathF.PI * newRot;
            float finalAngle = geo.EndAngle - MathF.PI + rotAngle * (newCW ? 1f : -1f);

            int endX = (int)MathF.Floor((geo.ArcCenterX + geo.ArcRadius * MathF.Cos(finalAngle) - geo.OffsetX) / geo.CellSize);
            int endY = (int)MathF.Floor((geo.ArcCenterY + geo.ArcRadius * MathF.Sin(finalAngle) - geo.OffsetY) / geo.CellSize);

            bool valid = grid.IsValid(endY, endX)
                      && grid.IsEmpty(endY, endX)
                      && (endX != (int)X || endY != (int)Y);

            if (!valid)
            {
                _angleOffset    = savedOffset;
                _arc            = savedArc;
                _clockwise      = savedCW;
                _rotationAmount = savedRot;
                return;
            }

            MoveTo(grid, config, endX, endY);
        }

        protected override List<GridPoint> GenLockPath(KGGrid grid, KGGridConfig config, int tx, int ty)
        {
            var geo    = ComputeGeometry(config);
            int steps  = Math.Max(10, (int)(_arc * _arc * (_rotationAmount * 8f)));
            float halfW   = config.CellSize * 0.5f - config.LineWidth * 0.51f;
            float rotAngle = 2f * MathF.PI * _rotationAmount;

            var path = new List<GridPoint>();
            var seen = new System.Collections.Generic.HashSet<GridPoint>();

            for (int i = 0; i <= steps; i++)
            {
                float t     = (float)i / steps;
                float angle = geo.EndAngle - MathF.PI + rotAngle * t * geo.RotationDirection;

                float cx = geo.ArcCenterX + geo.ArcRadius * MathF.Cos(angle);
                float cy = geo.ArcCenterY + geo.ArcRadius * MathF.Sin(angle);
                float innerR = MathF.Max(0.001f, geo.ArcRadius - halfW);
                float ix = geo.ArcCenterX + innerR * MathF.Cos(angle);
                float iy = geo.ArcCenterY + innerR * MathF.Sin(angle);
                float ox2 = geo.ArcCenterX + (geo.ArcRadius + halfW) * MathF.Cos(angle);
                float oy2 = geo.ArcCenterY + (geo.ArcRadius + halfW) * MathF.Sin(angle);

                foreach (var (wx, wy) in new[] { (cx, cy), (ix, iy), (ox2, oy2) })
                {
                    int gc = (int)MathF.Floor((wx - geo.OffsetX) / geo.CellSize);
                    int gr = (int)MathF.Floor((wy - geo.OffsetY) / geo.CellSize);

                    if (!grid.IsValid(gr, gc)) return null;

                    var pt = new GridPoint(gc, gr);
                    if (seen.Add(pt)) path.Add(pt);
                }
            }
            return path;
        }
    }
}

File: MacOS/Kinogrida/Classes/ArcShape/KGArcShape.swift
import AppKit

// ── Geometry bundle ───────────────────────────────────────────
private struct ArcGeometry {
    var centerX, centerY:     CGFloat
    var endAngle, initAngle:  CGFloat
    var startAngle, tailAngle: CGFloat
    var arcRadius:            CGFloat
    var arcCenterX, arcCenterY: CGFloat
    var cellSize, lineWidth:  CGFloat
    var rotationDirection:    CGFloat
    var offsetX, offsetY:     CGFloat
}

final class KGArcShape: KGBaseShape {

    var angleOffset: CGFloat    // radians
    var arc: CGFloat            // arc radius multiplier (in cells)
    var clockwise: Bool
    var rotationAmount: CGFloat // 0.25=quarter, 0.5=half, 0.75=three-quarter

    override init(grid: KGGrid, x: Int, y: Int, color: NSColor) {
        self.angleOffset    = KGArcShape.randomAngleOffset()
        self.arc            = 2
        self.clockwise      = true
        self.rotationAmount = 0.25
        super.init(grid: grid, x: x, y: y, color: color)
    }

    // ── Helpers ───────────────────────────────────────────────

    private static func randomAngleOffset() -> CGFloat {
        let angles: [CGFloat] = [0, 0.25, 0.5, 0.75]
        return angles[KGRandomInt(0, 3)] * .pi * 2
    }

    private func geometry(config: KGGridConfig, angleOffset: CGFloat, arc: CGFloat) -> ArcGeometry {
        let cs = config.cellSize
        let ox = config.offsetX, oy = config.offsetY
        let lw = config.lineWidth

        let centerX    = ox + x * cs + cs / 2
        let centerY    = oy + y * cs + cs / 2

        let endAngle   = angleOffset
        let initAngle  = endAngle - .pi

        let rotDir     = clockwise ? CGFloat(1) : -1
        let rotAngle   = 2 * CGFloat.pi * rotationAmount

        let startAngle = initAngle + rotAngle * progress     * rotDir
        let tailAngle  = initAngle + rotAngle * tailProgress * rotDir

        let arcRadius  = cs * arc
        let arcCenterX = centerX + cs * arc * cos(endAngle)
        let arcCenterY = centerY + cs * arc * sin(endAngle)

        return ArcGeometry(
            centerX: centerX,       centerY: centerY,
            endAngle: endAngle,     initAngle: initAngle,
            startAngle: startAngle, tailAngle: tailAngle,
            arcRadius: arcRadius,
            arcCenterX: arcCenterX, arcCenterY: arcCenterY,
            cellSize: cs,           lineWidth: lw,
            rotationDirection: rotDir,
            offsetX: ox,            offsetY: oy
        )
    }

    private func geometry(config: KGGridConfig) -> ArcGeometry {
        geometry(config: config, angleOffset: angleOffset, arc: arc)
    }

    // ── Draw arc elements (outer arc, inner arc, caps) ────────

    private func drawArcElements(ctx: CGContext, geo: ArcGeometry, color: NSColor, width: CGFloat) {
        let halfW = width / 2

        let rotAngle      = 2 * CGFloat.pi * rotationAmount
        let arcStartAngle = geo.initAngle + rotAngle * tailProgress * geo.rotationDirection
        let arcEndAngle   = geo.initAngle + rotAngle * progress     * geo.rotationDirection

        let startX = geo.arcCenterX + geo.arcRadius * cos(geo.startAngle)
        let startY = geo.arcCenterY + geo.arcRadius * sin(geo.startAngle)
        let endX   = geo.arcCenterX + geo.arcRadius * cos(geo.tailAngle)
        let endY   = geo.arcCenterY + geo.arcRadius * sin(geo.tailAngle)

        ctx.setLineWidth(geo.lineWidth)
        ctx.setStrokeColor(color.cgColor)

        // Start cap
        ctx.beginPath()
        ctx.addArc(center: CGPoint(x: startX, y: startY),
                   radius: max(0, halfW),
                   startAngle: clockwise ? geo.startAngle         : geo.startAngle - .pi,
                   endAngle:   clockwise ? geo.startAngle + .pi   : geo.startAngle,
                   clockwise: false)
        ctx.strokePath()

        // End cap
        ctx.beginPath()
        ctx.addArc(center: CGPoint(x: endX, y: endY),
                   radius: max(0, halfW),
                   startAngle: clockwise ? geo.tailAngle - .pi    : geo.tailAngle,
                   endAngle:   clockwise ? geo.tailAngle          : geo.tailAngle - .pi,
                   clockwise: false)
        ctx.strokePath()

        // Outer arc
        // Canvas uses counterclockwise = !self.clockwise. After Y-flip:
        // canvas CW (ccw=false) ↔ CGContext clockwise=false → !clockwise in Swift Bool
        ctx.beginPath()
        ctx.addArc(center: CGPoint(x: geo.arcCenterX, y: geo.arcCenterY),
                   radius: geo.arcRadius + halfW,
                   startAngle: arcStartAngle, endAngle: arcEndAngle,
                   clockwise: !clockwise)
        ctx.strokePath()

        // Inner arc
        ctx.beginPath()
        ctx.addArc(center: CGPoint(x: geo.arcCenterX, y: geo.arcCenterY),
                   radius: max(0.001, geo.arcRadius - halfW),
                   startAngle: arcStartAngle, endAngle: arcEndAngle,
                   clockwise: !clockwise)
        ctx.strokePath()
    }

    // ── Draw ──────────────────────────────────────────────────

    override func draw(in ctx: CGContext, config: KGGridConfig) {
        let geo      = geometry(config: config)
        let baseSize = config.cellSize * 0.8
        ctx.setLineCap(.round)
        drawArcElements(ctx: ctx, geo: geo, color: color, width: baseSize)
        drawArcElements(ctx: ctx, geo: geo, color: .white, width: baseSize - geo.lineWidth * 4)
    }

    // ── Update position (animation progress) ──────────────────
    // Progress update only — geometry finalization in updateProgress(config:grid:)

    override func updatePosition(grid: KGGrid) {
        guard isMoving else { return }

        if progress < 1 {
            progress = min(1, progress + speed)
            if progress >= 1 { hasReachedTarget = true }
        } else if tailProgress < 1 {
            tailProgress = min(1, tailProgress + speed)
        }
    }

    // Called by the engine after updateWithGrid to finalize arc grid position
    func updateProgress(config: KGGridConfig, grid: KGGrid) {
        guard isMoving, tailProgress >= 1 else { return }

        let geo        = geometry(config: config)
        let rotAngle   = 2 * CGFloat.pi * rotationAmount
        let finalAngle = geo.endAngle - .pi + rotAngle * geo.rotationDirection

        let newX = Int(floor((geo.arcCenterX + geo.arcRadius * cos(finalAngle) - geo.offsetX) / geo.cellSize))
        let newY = Int(floor((geo.arcCenterY + geo.arcRadius * sin(finalAngle) - geo.offsetY) / geo.cellSize))

        onMoveComplete(grid: grid, newX: newX, newY: newY)
        progress     = 0
        tailProgress = 0
    }

    // ── Calculate new target ──────────────────────────────────

    override func calculateNewTarget(grid: KGGrid, config: KGGridConfig) {
        guard !isMoving else { return }

        let maxArc = min(config.nbrColumns, config.nbrRows) / 2 - 1
        guard maxArc >= 1 else { return }

        let newOffset    = KGArcShape.randomAngleOffset()
        let newArc       = CGFloat(KGRandomInt(1, maxArc))
        let newClockwise = Bool.random()
        let rotAmounts: [CGFloat] = [0.25, 0.5, 0.75]
        let newRot       = rotAmounts[KGRandomInt(0, 2)]

        // Save current params
        let savedOffset = angleOffset, savedArc = arc
        let savedCW     = clockwise,   savedRot = rotationAmount

        // Temporarily apply candidate params to calculate endpoint
        angleOffset    = newOffset
        arc            = newArc
        clockwise      = newClockwise
        rotationAmount = newRot

        let geo        = geometry(config: config)
        let rotAngle   = 2 * CGFloat.pi * newRot
        let finalAngle = geo.endAngle - .pi + rotAngle * (newClockwise ? 1 : -1)

        let endX = Int(floor((geo.arcCenterX + geo.arcRadius * cos(finalAngle) - geo.offsetX) / geo.cellSize))
        let endY = Int(floor((geo.arcCenterY + geo.arcRadius * sin(finalAngle) - geo.offsetY) / geo.cellSize))

        let cellEmpty: Bool
        if endX >= 0 && endX < config.nbrColumns && endY >= 0 && endY < config.nbrRows {
            if case .empty = grid[endY, endX] { cellEmpty = true } else { cellEmpty = false }
        } else {
            cellEmpty = false
        }

        let valid = cellEmpty && (endX != Int(self.x) || endY != Int(self.y))

        guard valid else {
            angleOffset    = savedOffset
            arc            = savedArc
            clockwise      = savedCW
            rotationAmount = savedRot
            return
        }

        moveTo(grid: grid, config: config, x: endX, y: endY)
    }

    // ── Debug path ────────────────────────────────────────────

    func drawDebugPath(in ctx: CGContext, config: KGGridConfig) {
        guard isMoving else { return }

        let geo      = geometry(config: config)
        let steps    = max(10, Int(arc * arc * (rotationAmount * 8)))
        let halfW    = config.cellSize * 0.5 - config.lineWidth * 0.51
        let rotAngle = 2 * CGFloat.pi * rotationAmount

        var centerPts = [CGPoint]()
        var innerPts  = [CGPoint]()
        var outerPts  = [CGPoint]()

        for i in 0...steps {
            let t     = CGFloat(i) / CGFloat(steps)
            let angle = geo.endAngle - .pi + rotAngle * t * geo.rotationDirection
            let innerR = max(0.001, geo.arcRadius - halfW)

            centerPts.append(CGPoint(x: geo.arcCenterX + geo.arcRadius * cos(angle),
                                     y: geo.arcCenterY + geo.arcRadius * sin(angle)))
            innerPts.append(CGPoint(x: geo.arcCenterX + innerR * cos(angle),
                                    y: geo.arcCenterY + innerR * sin(angle)))
            outerPts.append(CGPoint(x: geo.arcCenterX + (geo.arcRadius + halfW) * cos(angle),
                                    y: geo.arcCenterY + (geo.arcRadius + halfW) * sin(angle)))
        }

        func drawPolyline(_ pts: [CGPoint], r: CGFloat, g: CGFloat, b: CGFloat) {
            ctx.setStrokeColor(red: r, green: g, blue: b, alpha: 1)
            ctx.setLineWidth(1)
            ctx.beginPath()
            for (i, p) in pts.enumerated() {
                if i == 0 { ctx.move(to: p) } else { ctx.addLine(to: p) }
                ctx.addArc(center: p, radius: 2, startAngle: 0, endAngle: .pi * 2, clockwise: false)
                ctx.move(to: p)
            }
            ctx.strokePath()
        }

        drawPolyline(centerPts, r: 1, g: 0, b: 0) // red
        drawPolyline(innerPts,  r: 0, g: 0, b: 1) // blue
        drawPolyline(outerPts,  r: 0, g: 1, b: 0) // green
    }

    // ── Lock path (arc cells) ─────────────────────────────────

    override func genLockPath(grid: KGGrid, config: KGGridConfig, targetX: Int, targetY: Int) -> [GridPoint]? {
        let geo      = geometry(config: config)
        let steps    = max(10, Int(arc * arc * (rotationAmount * 8)))
        let halfW    = config.cellSize * 0.5 - config.lineWidth * 0.51
        let rotAngle = 2 * CGFloat.pi * rotationAmount

        var path = [GridPoint]()
        var seen = Set<GridPoint>()

        for i in 0...steps {
            let t     = CGFloat(i) / CGFloat(steps)
            let angle = geo.endAngle - .pi + rotAngle * t * geo.rotationDirection

            let cx     = geo.arcCenterX + geo.arcRadius * cos(angle)
            let cy     = geo.arcCenterY + geo.arcRadius * sin(angle)
            let innerR = max(0.001, geo.arcRadius - halfW)
            let ix     = geo.arcCenterX + innerR * cos(angle)
            let iy     = geo.arcCenterY + innerR * sin(angle)
            let ox2    = geo.arcCenterX + (geo.arcRadius + halfW) * cos(angle)
            let oy2    = geo.arcCenterY + (geo.arcRadius + halfW) * sin(angle)

            for (wx, wy) in [(cx, cy), (ix, iy), (ox2, oy2)] {
                let gc = Int(floor((wx - geo.offsetX) / geo.cellSize))
                let gr = Int(floor((wy - geo.offsetY) / geo.cellSize))

                guard grid.isValid(row: gr, col: gc) else { return nil }

                let pt = GridPoint(col: gc, row: gr)
                if seen.insert(pt).inserted {
                    path.append(pt)
                }
            }
        }
        return path
    }
}
